src/main/java/com/google/devtools/build/lib/skyframe/BzlmodRepoRuleFunction.java

```diff
@@ -0,0 +1,256 @@
+
+    // @bazel_tools is a special repo that we pull from the extracted install dir.
+    if (repositoryName.equals(TOOLS_REPO)) {
+      repoSpec = new RepoSpec(
+          "local_repository",
+          ImmutableMap.of(
+              "name", "bazel_tools", "path",
+              directories.getEmbeddedBinariesRoot().getChild("embedded_tools").getPathString()));
+      return getRuleFromSpec(repoSpec, starlarkSemantics, env);
+    }
+
+    // Look for repositories defined by non-registry overrides.
+    RepoSpecsValue overrideDepRepos =
+        (RepoSpecsValue) env.getValue(RepoSpecsValue.KEY_FOR_OVERRIDE_DEP);
+    if (overrideDepRepos == null) {
+      return null;
+    }
+    repoSpec = overrideDepRepos.getRepository(repositoryName);
+    if (repoSpec != null) {
+      return getRuleFromSpec(repoSpec, starlarkSemantics, env);
+    }
+
+    // Look for repositories derived from native Bazel Modules
+    RepoSpecsValue bazelModuleRepos =
+        (RepoSpecsValue) env.getValue(RepoSpecsValue.KEY_FOR_BAZEL_MODULE);
+    if (bazelModuleRepos == null) {
+      return null;
+    }
+    repoSpec = bazelModuleRepos.getRepository(repositoryName);
+    if (repoSpec != null) {
+      return getRuleFromSpec(repoSpec, starlarkSemantics, env);
+    }
+
+    // Look for repositories derived from module rules if the repo is not requested for module rule
+    // resolution.
+    RepoSpecsValue moduleRuleRepos =
+        (RepoSpecsValue) env.getValue(RepoSpecsValue.KEY_FOR_MODULE_RULE);
+    if (moduleRuleRepos == null) {
+      return null;
+    }
+    repoSpec = moduleRuleRepos.getRepository(repositoryName);
+    if (repoSpec != null) {
+      return getRuleFromSpec(repoSpec, starlarkSemantics, env);
+    }
+
+    return BzlmodRepoRuleValue.REPO_RULE_NOT_FOUND_VALUE;
+  }
+
+  private BzlmodRepoRuleValue getRuleFromSpec(RepoSpec repoSpec,
+      StarlarkSemantics starlarkSemantics, Environment env)
+      throws BzlmodRepoRuleFunctionException, InterruptedException {
+    if (isNativeRepoRule(repoSpec)) {
+      return getNativeRepoRule(repoSpec, starlarkSemantics, env);
+    } else {
+      return getStarlarkRepoRule(repoSpec, starlarkSemantics, env);
+    }
+  }
+
+  private boolean isNativeRepoRule(RepoSpec info) {
+    // For native repo rule, the rule class name is just <rule class name>;
+    // For Starlark repo rule, the rule class name is <label for bzl file>%<rule class name>
+    // eg. Native: local_repository
+    //     Starlark: //:repo.bzl%my_repo
+    return !info.getRuleClass().contains("%");
+  }
+
+  // Create the external package builder, which is only for the convenience of creating
+  // repository rules.
+  private Package.Builder getExternalPackageBuilder(StarlarkSemantics semantics) {
+    RootedPath bzlmodFile =
+        RootedPath.toRootedPath(
+            Root.fromPath(directories.getWorkspace()), LabelConstants.MODULE_DOT_BAZEL_FILE_NAME);
+
+    Package.Builder pkg =
+        packageFactory.newExternalPackageBuilder(
```

Comment:
Given that using `PackageBuilder` for repositories is really an antipattern since it forces us to add data specific to repository handling to `Package`, I'd rather we don't use it here. Looking at `RepositoryDelegatorFunction.getHandler()`, all this is used for is to get the correct `RepositoryFunction`, which looks achievable without getting `Rule` involved.